scd/token.c

```diff
@@ -0,0 +1,321 @@
+int
+int_change_pw_st(scd_token_t *token, const char *oldpass, const char *newpass,
+		 UNUSED unsigned char *pairing_secret, UNUSED size_t pairing_sec_len,
+		 UNUSED bool is_provisioning)
+{
+	return softtoken_change_passphrase(token->token_data->int_token.softtoken, oldpass,
+					   newpass);
+}
+
+/*  -----------------------------------------------------------------------  */
+#ifdef ENABLESCHSM
+
+int
+int_lock_usb(scd_token_t *token)
+{
+	return usbtoken_lock(token->token_data->int_token.usbtoken);
+}
+
+int
+int_unlock_usb(scd_token_t *token, char *passwd, unsigned char *pairing_secret,
+	       size_t pairing_sec_len)
+{
+	TRACE("SCD: int_usb_unlock");
+	return usbtoken_unlock(token->token_data->int_token.usbtoken, passwd, pairing_secret,
+			       pairing_sec_len);
+}
+
+bool
+int_is_locked_usb(scd_token_t *token)
+{
+	return usbtoken_is_locked(token->token_data->int_token.usbtoken);
+}
+
+bool
+int_is_locked_till_reboot_usb(scd_token_t *token)
+{
+	return usbtoken_is_locked_till_reboot(token->token_data->int_token.usbtoken);
+}
+
+int
+int_wrap_usb(scd_token_t *token, char *label, unsigned char *plain_key, size_t plain_key_len,
+	     unsigned char **wrapped_key, int *wrapped_key_len)
+{
+	return usbtoken_wrap_key(token->token_data->int_token.usbtoken, (unsigned char *)label,
+				 strlen(label), plain_key, plain_key_len, wrapped_key,
+				 wrapped_key_len);
+}
+
+int
+int_unwrap_usb(scd_token_t *token, char *label, unsigned char *wrapped_key, size_t wrapped_key_len,
+	       unsigned char **plain_key, int *plain_key_len)
+{
+	return usbtoken_unwrap_key(token->token_data->int_token.usbtoken, (unsigned char *)label,
+				   strlen(label), wrapped_key, wrapped_key_len, plain_key,
+				   plain_key_len);
+}
+
+int
+int_change_pw_usb(scd_token_t *token, const char *oldpass, const char *newpass,
+		  unsigned char *pairing_secret, size_t pairing_sec_len, bool is_provisioning)
+{
+	return usbtoken_change_passphrase(token->token_data->int_token.usbtoken, oldpass, newpass,
+					  pairing_secret, pairing_sec_len, is_provisioning);
+}
+#endif // ENABLESCHSM
+
+/*  -----------------------------------------------------------------------  */
+
+/**
+ * creates a new generic token
+ * calls the respective create function for the selected type of token and
+ * sets the function pointer appropriately
+ */
+scd_token_t *
+token_new(token_constr_data_t *constr_data)
+{
+	scd_token_t *new_token;
+	char *token_file = NULL;
+
+	TRACE("SCD: scd_token_new");
+
+	new_token = mem_new0(scd_token_t, 1);
+	if (!new_token) {
+		ERROR("Could not allocate new scd_token_t");
+		return NULL;
+	}
+
+	new_token->token_data = mem_new(scd_token_data_t, 1);
+	if (!new_token->token_data) {
+		ERROR("Could not allocate memory for token_data_t");
+		goto err_token_data;
+	}
+
+	new_token->token_data->token_uuid = uuid_new(constr_data->uuid);
+	if (!new_token->token_data->token_uuid) {
+		ERROR("Could not allocate memory for token_uuid");
+		goto err_token_uuid;
```

Comment:
define a single error target and check with if (var) {free(var)},
see also below

Reply:
done